src/cockpit/channels/filesystem.py

```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
-                raise ChannelError('not-found', message=str(exc)) from exc
-            except IsADirectoryError as exc:
-                self.unlink_temppath()
-                # not ideal, but the closest code we have
-                raise ChannelError('access-denied', message=str(exc)) from exc
-            except OSError as exc:
-                self.unlink_temppath()
-                raise ChannelError('internal-error', message=str(exc)) from exc
-
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
+                    loop = asyncio.get_running_loop()
+                    while data is not None:
+                        await loop.run_in_executor(None, tmp.write, data)
+                        data = await self.read()
+
+                    os.fdatasync(tmp.fileno())
+
+                    if tag is None:
+                        # no preconditions about what currently exists or not
+                        # calculate the file mode from the umask
+                        os.fchmod(tmp.fileno(), 0o666 & ~my_umask())
+                        os.rename(tmp.name, path)
+                        tmp = None  # don't need to delete
+
+                    elif tag == '-':
+                        # the file must not exist.  file mode from umask.
+                        os.fchmod(tmp.fileno(), 0o666 & ~my_umask())
+                        os.link(tmp.name, path)  # will fail if file exists
+
+                    else:
+                        # the file must exist with the given tag
+                        buf = os.stat(path)
+                        if tag != tag_from_stat(buf):
+                            raise ChannelError('change-conflict')
+                        # chown/chmod from the existing file permissions
+                        os.fchmod(tmp.fileno(), stat.S_IMODE(buf.st_mode))
+                        os.fchown(tmp.fileno(), buf.st_uid, buf.st_gid)
+                        os.rename(tmp.name, path)
+                        tmp = None  # don't need to delete
+
+        except FileNotFoundError as exc:
+            raise ChannelError('not-found') from exc
+        except FileExistsError as exc:
+            # that's from link() noticing that the target file already exists
+            raise ChannelError('change-conflict') from exc
+        except PermissionError as exc:
+            raise ChannelError('access-denied') from exc
+        except IsADirectoryError as exc:
+            # not ideal, but the closest code we have
+            raise ChannelError('access-denied', message=str(exc)) from exc
+        except OSError as exc:
+            raise ChannelError('internal-error', message=str(exc)) from exc
+        finally:
+            if tmp is not None:
+                os.unlink(tmp.name)
```

Comment:
See above -- doing that should be the whole point of using a `with` context? I seriously doubt my Python knowledge right now..

Reply:
We create the `NamedTempFile()` with `delete=True`, which means that the contextmanager will close the tempfile, but not delete it.  This would all be so much better if we could use `delete_on_close=True`, but it's only in 3.12 :(

Reply:
IMHO `delete_on_close=True` is the whole reason why we want to use NamedTemporaryFile. I.e. *always* remove the temp file, and just rename it to the final name in the happy path. I'd consider a rewrite using `=False` as making it worse..

Reply:
From where I'm sitting: context managers are nice, but so are `finally:` clauses.  They're essentially interchangeable.

`NamedTemporaryFile`, in the best case, does ~4 things:
 - creates a unique name
 - gives you a file-like object for interacting with it
 - closes it when you're done
 - deletes it when you're done

The last one is not working for us, so we have to go manual on that one.... but I'm happy to use the other 3...

Reply:
Thanks for evaluating all the options. So the summary is "they all suck, Linux doesn't have a good API for this, let's just pick the least bad one"  :sweat_smile: 

So, I'm ok with this, we just need to figure out the fsreplace/fswatch race below.